src/spaces/continuous.jl

```diff
@@ -410,3 +410,230 @@ function Base.iterate(iter::PairIterator, i = 1)
+
+#######################################################################################
+# FMP Algorithm
+#######################################################################################
+export fmp_update_vel, fmp_update_interacting_pairs, FMP_Parameters, fmp_parameter_init
+
+"""
+    FMP_Parameters
+The parameters for the FMP model as defined in the FMP paper. Helper function `FMP_Parameter_Init`
+is used to initialize the struct with default parameters. Users can modify FMP default parameters
+through the usage of keyword arguments. See [the original paper](https://arxiv.org/abs/1909.05415)
+for a complete description of model parameters. Parameters included in this struct that are not 
+present in the original paper are:
+- `obstacle_list`: list of obstacles in the state space
+- `interaction_array`: n x n boolean array (n = number of agents) where `interaction_array[i,j]=1`
+indicates that the ith and jth agents are interacting. Used so that `interacting_pairs` is only called
+once per iteration rather than every time `update_vel` is called.
+- `agents`: an iterator of the interacting agents from `interacting_pairs` at each time step.
+"""
+mutable struct FMP_Parameters
+    rho::Float64
+    rho_obstacle::Float64
+    step_inc::Int64
+    terminal_max_dis::Float64
+    c1::Float64
+    c2::Float64
+    vmax::Float64
+    d::Float64
+    r::Float64
+    obstacle_list::Array{Int64}
+    interaction_array::Array{Bool}
+    agents::Dict{Int64, Any}
+    FMP_Parameters() = new()
+end
+
+"""
+    fmp_update_interacting_pairs(model)
+Updates `FMP_Parameters.interaction_array` with the current array of interacting agents by calling
+`interacting_pairs`. It does this once per model step to reduce the potential overhead of calling
+`interacting_pairs` each time `update_vel` is called. 
+"""
+function fmp_update_interacting_pairs(
+    model::ABM{<:ContinuousSpace},
+    )
+
+    # get list of interacting_pairs within some radius
+    agent_iter = interacting_pairs(model, model.FMP_params.r, :all)
+    for agent_id in keys(model.agents)
+        Ni = Int64[]
+        Gi = Int64[]
+        for (i,j) in agent_iter.pairs
+            # handle agent-agent interactions
+            if i == agent_id && model.agents[j].type == :A
+                append!(Ni, j)
+            elseif j == agent_id && model.agents[i].type == :A
+                append!(Ni, i)
+            
+            # handle agent-goal interactions
+            elseif i == agent_id && model.agents[j].type == :T
+                append!(Gi, j)
+            elseif j == agent_id && model.agents[i].type == :T
+                append!(Gi, i)
+            end
+        end
+
+        model.agents[agent_id].Ni = Ni
+        model.agents[agent_id].Gi = Gi
+    end
+
+end
+
+"""
+    fmp_parameter_init()
+A convenience function for initializing the FMP_Parameters struct with typical FMP
+parameters. Users can modify the FMP algorithm parameters through the usage of 
+keyword arguments. 
+"""
```

Comment:
It *truly doesn't make sense* to have this function *without* stating what keywords it accepts..........